[boot] Add LOCALIP= and getty= to /bootopts

ghaerr opened the pull request:
Adds two items requested in #1155.

LOCALIP= specifies ktcp local IP address, and works whenever `net start` is run.

getty=19200 (or other baud rate) overrides any /bin/getty line in /etc/inittab that also specifies a baud rate. This allows /etc/inittab to remain unchanged, yet be overidden at boot when desired.
/etc/net.cfg defaults to not running httpd for now.

Also fixes ":once:" and non-/bin/getty commands, they were broken, now run correctly using "sh -c" to run from init.
Various other compilation warnings addressed.

ghaerr committed:
[boot] Add LOCALIP= and getty= to /bootopts

Mellvik commented:
Thanks again, @ghaerr - 

this is fast turnaround indeed.

The LOCALIP= is exactly what's needed.

As to the serial settings I'm a more difficult customer. We still need to make two mods in order for this to work:

## boot options max size 511 bytes
console=ttyS0,57600 debug net=eth 3 # serial console, multiuser, networking
#QEMU=1                 # to use ftp/ftpd in qemu
#TZ=MDT7
#LOCALIP=10.0.2.16
#netirq=9 netport=0x300 # NIC
bufs=2500               # system buffers
sync=30         # autosync secs
#init=/bin/init 3 n     # multiuser serial no /etc/rc.sys
#init=/bin/sh           # singleuser shell
getty=57600

It's an improvement since only one file needs to be edited, but to me the addition of another variable seems like a complication that can be avoided. 
Maybe I was ambiguous in my description of the problem. My thinking was that init would pick up the speed setting from the console= line, thus connect the serial port and the speed setting, and use that to override that particular getty. From my perspective it's advantageous if this mod only affects the console, not the other serial lines.

And BTW thanks for the net.cfg mod - appreciated.

—M


> 15. feb. 2022 kl. 01:45 skrev Gregory Haerr ***@***.***>:
> 
> Adds two items requested in #1155 <https://github.com/jbruchon/elks/issues/1155>.
> 
> LOCALIP= specifies ktcp local IP address, and works whenever net start is run.
> 
> getty=19200 (or other baud rate) overrides any /bin/getty line in /etc/inittab that also specifies a baud rate. This allows /etc/inittab to remain unchanged, yet be overidden at boot when desired.
> /etc/net.cfg defaults to not running httpd for now.
> 
> Also fixes ":once:" and non-/bin/getty commands, they were broken, now run correctly using "sh -c" to run from init.
> Various other compilation warnings addressed.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/jbruchon/elks/pull/1157 <https://github.com/jbruchon/elks/pull/1157>
> Commit Summary
> 
> 374cc93 <https://github.com/jbruchon/elks/pull/1157/commits/374cc934e05c694aa27ae1233dc68ba14db709e0> [boot] Add LOCALIP= and getty= to /bootopts
> File Changes (6 files <https://github.com/jbruchon/elks/pull/1157/files>)
> M elkscmd/rootfs_template/bootopts <https://github.com/jbruchon/elks/pull/1157/files#diff-2c59e1666f07f8333ccf7ad2885be1d2e2100ca14e30558ff71a87e770542cf3> (8)
> M elkscmd/rootfs_template/etc/inittab <https://github.com/jbruchon/elks/pull/1157/files#diff-7f9bf2131132af5c5b04baa77f48e2aba9ff7a2a3c3bafac9ea7b71873383c4e> (2)
> M elkscmd/rootfs_template/etc/net.cfg <https://github.com/jbruchon/elks/pull/1157/files#diff-09c55d1cded5ed30fed005cf5e52462309064ad3d28e9d1820509c7a49cb7e1b> (7)
> M elkscmd/sys_utils/init.c <https://github.com/jbruchon/elks/pull/1157/files#diff-45ea2e9eacd2393b6f52784d90b1f1a5eb44dd890c72e56d6ea820f197b8a525> (21)
> M libc/getent/__getpwent.c <https://github.com/jbruchon/elks/pull/1157/files#diff-61947b1830672d51fdfc5f144ac93c2e55ce701647b5d244fa6ce9655a82bd5e> (1)
> M libc/system/signal.c <https://github.com/jbruchon/elks/pull/1157/files#diff-45bc7868b34c5017bb98b60b4d6ed92c9dbfd8310812835fd455c9c749b56006> (1)
> Patch Links:
> 
> https://github.com/jbruchon/elks/pull/1157.patch <https://github.com/jbruchon/elks/pull/1157.patch>
> https://github.com/jbruchon/elks/pull/1157.diff <https://github.com/jbruchon/elks/pull/1157.diff>
> —
> Reply to this email directly, view it on GitHub <https://github.com/jbruchon/elks/pull/1157>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AA3WGOCNINWSDJE7GLBXDMTU3GO2FANCNFSM5ONAGPOQ>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
> You are receiving this because you are subscribed to this thread.
>

ghaerr commented:
> Maybe I was ambiguous in my description of the problem.

Yes, I didn't understand until your last post and more precise description. 

>  From my perspective it's advantageous if this mod only affects the console, not the other serial lines.

I thought you were saying you needed something that would work for two serial ports. But I think I understand now what you need.

> t's an improvement since only one file needs to be edited, but to me the addition of another variable seems like a complication that can be avoided.

Yes, agree that a 'getty=' line is kind of a hack, now that I think more about it.

> My thinking was that init would pick up the speed setting from the console= line, thus connect the serial port and the speed setting, and use that to override that particular getty. 

We can't use the console= line to pass any data, since that's not an environment variable and that information can't be seen at /bin/init time. Only the env=value lines do that. However, one possible solution would be that the serial port remembers it's last value (or the console= value) on subsequent opens. It may already be doing that, but certainly the /etc/inittab getty lines will override any previous setting. That could maybe be fixed be removing the baud rate from the /etc/inittab line(s).

Overall, still complicated. How about something much simpler? Remove the getty= change, and just ship /bootopts with a console=ttyS0,57600 already set (commented out), and also change the /etc/inittab getty lines to 57600 default, for now?

I have no problem changing to 57600 default in both places, to aid your ability to continue to deploy and test! We can change them to other defaults later, if desired, but 57600 is a more modern serial speed anyways. Also, funnily enough, this issue resulted in finding (yet another) bug in /bin/init - all non-/bin/getty lines failed to function!

Mellvik commented:
> 
> Overall, still complicated. How about something much simpler? Remove the getty= change, and just ship /bootopts with a console=ttyS0,57600 already set (commented out), and also change the /etc/inittab getty lines to 57600 default, for now?
> 
> I have no problem changing to 57600 default in both places, to aid your ability to continue to deploy and test! We can change them to other defaults later, if desired, but 57600 is a more modern serial speed anyways. Also, funnily enough, this issue resulted in finding (yet another) bug in /bin/init - all non-/bin/getty lines failed to function!
> 
> 
That would certainly do it for now, @ghearr - no problem, appreciated.

While we're at it, and since you mention it: If the tty remembers its speed setting from the console= (likely since it's outputting at that rate during the boot sequence), and since there is an overall system default speed of 9600, it might be possible to have init check if the speed is different (than 9600) on the console port, and override inittab with that speed when starting getty?

Wow, that was a long sentence. Was that intelligible at all?

Thank you!

—M

ghaerr commented:
> That would certainly do it for now, @ghaerr - no problem, appreciated.

I have this done, except for one item: do you want only ttyS0 set to 57600, or also ttyS1 to 57600 in /etc/inittab? I have changed console=ttyS0,57600 in /bootopts and /bin/getty ttyS0 57600 in /etc/inittab.

>  it might be possible to have init check if the speed is different (than 9600) on the console port, and override inittab with that speed when starting getty?

I understand your request. That means there would be some cross-checking between ports in the serial drive and silently changing a baud rate. I feel we're better off keeping things simple, and keeping the ability that only the serial ports specified (anywhere) have their values changed. This will certainly help when there's a baud rate connection issue, which we know does happen.

I have also added a `net show` option that displays the settings that would/do take place when a `net start` is executed. This will help you quickly see everything, except for the case when specific parameters are passed to net start. You can play with that and see what you think.

ghaerr commented:
> it might be possible to have init check if the speed is different (than 9600) on the console port, and override inittab with that speed when starting getty?

I might have interpreted this wrong. I think if we're going to do something tricky, it should be done only when there is NO baud rate specified in the /bin/getty line(s) in /etc/inittab. In that way, /bin/init (nor getty) should not change the baud rate at all, but use the previous setting... that is if ELKS currently doesn't reset the baud rate between open/closes. Perhaps you could check that and we'll talk about it after this issue set of changes is committed.

ghaerr commented:
I still don't know exactly what you're trying to do. Do you want ttyS1 set to a baud rate always specified explicitly in /etc/inittab, but have the ability to change ttyS0 in /bootopts, and also have that setting work for logins on ttyS0? Or something else?

ghaerr commented:
I just checked the source for getty.c as well as the serial driver. It seems the serial driver will keep a previous baud rate, which is initially either set to the system default (9600), or by /bootopts. If a baud rate is NOT specified in /etc/inittab, then getty does NOT reset the baud rate. So I think your problem is solved by leaving the ttyS0 line with no baud rate in /etc/inittab, and specify it in /bootopts. The ttyS1 baud rate will be set to 9600 at boot, and then reset to whatever value is in /etc/inittab. 

Also, if a /etc/inittab getty spawns a login and the user changes the baud rate, logging out and then in again will not change the baud rate. I think this is what we want, for now.

I just need to know  what to set the /etc/inittab lines for ttyS0 and ttyS1 to. I presume none for ttyS0 and need to know what ttyS1 should be set to by default.

Mellvik commented:
Agreed, @ghaerr,

let's forget about ttyS1 altogether (IOW keep it as is). It will be a very odd case when console is serial and not ttyS0.

My recent mentioning of ttyS1 was in the context of the general getty= in bootopts, which would set the ports to the same speed. That would not be desirable because they have different uses (the 2nd line more often used for file transfers).

so yes, lets stick to ttyS0, and the problem to solve is simply to ensure that the boot messages and the getty login prompt end up with the same baud rate.

And yes, I'll check whether ELKS actually resets speed between opens/closes. It's easily down with stty.


—M



> 15. feb. 2022 kl. 18:48 skrev Gregory Haerr ***@***.***>:
> 
> 
> I still don't know exactly what you're trying to do. Do you want ttyS1 set to a baud rate always specified explicitly in /etc/inittab, but have the ability to change ttyS0 in /bootopts, and also have that setting work for logins on ttyS0? Or something else?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/jbruchon/elks/pull/1157#issuecomment-1040589173>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AA3WGOG5GJWC32DU3EEQSMLU3KGVPANCNFSM5ONAGPOQ>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
> You are receiving this because you commented.
>

Mellvik commented:
COnfirmed: The speed sticks between opens. 

So if getty leaves the speed alone if missing in inittab, we should be fine. Actually OK regardless since then the speed will be 9600 if nothing is changed in /booptopts.

Thank you.

—M


> 15. feb. 2022 kl. 18:57 skrev Helge Skrivervik ***@***.***>:
> 
> Agreed, @ghaerr,
> 
> let's forget about ttyS1 altogether (IOW keep it as is). It will be a very odd case when console is serial and not ttyS0.
> 
> My recent mentioning of ttyS1 was in the context of the general getty= in bootopts, which would set the ports to the same speed. That would not be desirable because they have different uses (the 2nd line more often used for file transfers).
> 
> so yes, lets stick to ttyS0, and the problem to solve is simply to ensure that the boot messages and the getty login prompt end up with the same baud rate.
> 
> And yes, I'll check whether ELKS actually resets speed between opens/closes. It's easily down with stty.
> 
> 
> —M
> 
> 
> 
>> 15. feb. 2022 kl. 18:48 skrev Gregory Haerr ***@***.*** ***@***.***>>:
>> 
>> 
>> I still don't know exactly what you're trying to do. Do you want ttyS1 set to a baud rate always specified explicitly in /etc/inittab, but have the ability to change ttyS0 in /bootopts, and also have that setting work for logins on ttyS0? Or something else?
>> 
>> —
>> Reply to this email directly, view it on GitHub <https://github.com/jbruchon/elks/pull/1157#issuecomment-1040589173>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AA3WGOG5GJWC32DU3EEQSMLU3KGVPANCNFSM5ONAGPOQ>.
>> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
>> You are receiving this because you commented.
>> 
>

ghaerr commented:
> so yes, lets stick to ttyS0, and the problem to solve is simply to ensure that the boot messages and the getty login prompt end up with the same baud rate.

> So if getty leaves the speed alone if missing in inittab, we should be fine. Actually OK regardless since then the speed will be 9600 if nothing is changed in /booptopts.

Ok, I'll post this PR for your inspection. It will remove the baud rate from /etc/inittab for ttyS0, and should work. If it doesn't we'll fix it so that it does. This means that /bootopts will control the startup speed ttyS0, using the console= line.

ghaerr committed:
Remove getty= /bootopts option, remove forced 9600 baud in /etc/inittab for ttyS0

ghaerr commented:
> COnfirmed: The speed sticks between opens.

Great. I will commit this now.  At this time, I am not thinking of making further changes to mount.cfg discussed in #1155.

I suggest you test this, and then add some specific script code in mount.cfg that seems to work well (for fsck etc), and post that in #1155 for further discussion on what to do, if anything. I can then parameterize it etc for final distribution.
